### PR TITLE
Feat!: Change TaskDef hinting system

### PR DESCRIPTION
**Pull request by SebastianMorawiec**
# The problem
TaskDef had flawed hinting system, leading to false-positives in Pyright.

# This PR's solution
Change our hinting system in TaskDef.
https://zapatacomputing.atlassian.net/wiki/spaces/ORQSRUN/pages/851869699/2024-01-09+Type+Hinting+of+DSL+TaskDef+object for more explanation

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/imp[rovement]/int[ernal]/docs>[!]:`
- [x] The PR is linked to a JIRA ticket (if there's no suitable ticket, check the box).

### CODE REVIEW

File: src/orquestra/sdk/_base/_dsl.py [hunk 1]
```diff
@@ -818,19 +806,19 @@ def with_invocation_meta(
         self,
         *,
         cpu: Optional[
-            Union[str, "orquestra.sdk._base._dsl.Sentinel"]
+            Union[str, "orquestra.sdk._base._dsl.Sentinel"]  # pyright: ignore
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_dsl.py:809]: **non-blocking**: what's the error raised by pyright here?

File: src/orquestra/sdk/_base/_dsl.py [hunk 2]
```diff
@@ -745,10 +733,10 @@ class ArtifactFuture:
 
     def __init__(
         self,
-        invocation: orquestra.sdk._base._dsl.TaskInvocation,
+        invocation: orquestra.sdk._base._dsl.TaskInvocation,  # pyright: ignore
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_dsl.py:736]: **non-blocking**: what's the error raised by pyright here?

File: src/orquestra/sdk/_base/_dsl.py [hunk 3]
```diff
@@ -745,10 +733,10 @@ class ArtifactFuture:
...
         output_index: Optional[int] = None,
         custom_name: Optional[str] = DEFAULT_CUSTOM_NAME,
-        serialization_format: orquestra.sdk._base._dsl.ArtifactFormat = DEFAULT_SERIALIZATION_FORMAT,  # noqa: E501
+        serialization_format: orquestra.sdk._base._dsl.ArtifactFormat = DEFAULT_SERIALIZATION_FORMAT,  # noqa: E501   # pyright: ignore
```
Comment (alexjuda) [inline src/orquestra/sdk/_base/_dsl.py:739]: **non-blocking**: what's the error raised by pyright here?

**Review by alexjuda — CHANGES_REQUESTED** [review]
LGTM, but we want to give it another view before merging [discussed on Zoom]

**Review by jamesclark-Zapata — COMMENTED** [review]
To me, this is a breaking change because type hints can no longer be relied upon.
Please make sure you put this in the breaking changes in the changelog.

A minor change to that ADR (sorry for not reviewing) is to make `ArtifactFuture` generic, then we can at least display the original type in the user's IDE:
Screenshot 2024-03-19 at 18 07 19

See my branch on top of this one: https://github.com/zapatacomputing/orquestra-workflow-sdk/tree/tmp/jamesclark-zapata/generic-artifacts

File: src/orquestra/sdk/_base/_dsl.py (hunk 3 above)
Comment (SebastianMorawiec) [reply src/orquestra/sdk/_base/_dsl.py]: For all of the answers: pyright doesnt like the _dsl import, which it is right. but we keep the fully-qualified types for docs apperently (at least thats what is stated in the comment)

**Review by alexjuda — APPROVED** [review]
🔥

**Pull request merged**